flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/source/enumerator/ContinuousSplitPlannerImpl.java

```diff
@@ -0,0 +1,160 @@
+    this.workerPool = ThreadPools.newWorkerPool(
+        "iceberg-plan-worker-pool-" + threadPoolName, scanContext.planParallelism());
+  }
+
+  @Override
+  public ContinuousEnumerationResult planSplits(IcebergEnumeratorPosition lastPosition) {
+    table.refresh();
+    if (lastPosition != null) {
+      return discoverIncrementalSplits(lastPosition);
+    } else {
+      return discoverInitialSplits();
+    }
+  }
+
+  /**
+   * Discover incremental changes between @{code lastPosition} and current table snapshot
+   */
+  private ContinuousEnumerationResult discoverIncrementalSplits(IcebergEnumeratorPosition lastPosition) {
+    // incremental discovery mode
+    Snapshot currentSnapshot = table.currentSnapshot();
+    if (currentSnapshot.snapshotId() == lastPosition.endSnapshotId()) {
+      LOG.info("Current table snapshot is already enumerated: {}", currentSnapshot.snapshotId());
+      return new ContinuousEnumerationResult(Collections.emptyList(), lastPosition);
+    } else {
+      ScanContext incrementalScan = scanContext
+          .copyWithAppendsBetween(lastPosition.endSnapshotId(), currentSnapshot.snapshotId());
+      LOG.info("Incremental scan: startSnapshotId = {}, endSnapshotId = {}",
+          incrementalScan.startSnapshotId(), incrementalScan.endSnapshotId());
+      List<IcebergSourceSplit> splits = FlinkSplitPlanner.planIcebergSourceSplits(table, incrementalScan, workerPool);
+      IcebergEnumeratorPosition position = IcebergEnumeratorPosition.builder()
+          .startSnapshotId(lastPosition.endSnapshotId())
+          .startSnapshotTimestampMs(lastPosition.endSnapshotTimestampMs())
+          .endSnapshotId(currentSnapshot.snapshotId())
+          .endSnapshotTimestampMs(currentSnapshot.timestampMillis())
+          .build();
+      LOG.info("Discovered {} splits from incremental scan: {}", splits.size(), position);
+      return new ContinuousEnumerationResult(splits, position);
+    }
+  }
+
+  /**
+   * Discovery initial set of splits based on {@link StreamingStartingStrategy}.
+   *
+   * <li>{@link ContinuousEnumerationResult#splits()} should contain initial splits
+   * discovered from table scan for {@link StreamingStartingStrategy#TABLE_SCAN_THEN_INCREMENTAL}.
+   * For all other strategies, splits collection should be empty.
+   * <li>{@link ContinuousEnumerationResult#position()} points to the starting position
+   * for the next incremental split discovery with exclusive behavior. Meaning files committed
+   * by the snapshot from the position in {@code ContinuousEnumerationResult} won't be included
+   * in the next incremental scan.
+   */
+  private ContinuousEnumerationResult discoverInitialSplits() {
+    Optional<Snapshot> startSnapshotOptional = getStartSnapshot(table, scanContext);
+    if (!startSnapshotOptional.isPresent()) {
+      // set IcebergEnumeratorPosition to null in this case
+      return new ContinuousEnumerationResult(Collections.emptyList(), null);
+    }
```

Comment:
guess we can not return null position here, then the next plan will still invoke `#discoverInitialSplits`. If some snapshots produced during this period， they will be seen as start snapshots and excluded.

Reply:
Empty table is invalid if the strategies must correspond to a certain snapshot, but it's weird to say empty table can not be applied with these two strategies.

There may be more than one snapshots can be committed before next round of discover, so it's not only the first snapshot.

How about returning a position with end snapshot as -1, which i think is idiomatic in iceberg.

Reply:
agree it is not just the first snapshot. I was using it just to simplify the discussion. it doesn't really matter how many snapshots are committed before the first discovery. semantics are the same.

I was saying `earliest` or `latest` snapshot can be interpreted as the earliest or latest valid snapshot of the table. Let's say when the Flink job started, it is an empty table. initial scan return null. 
* case 1: there are 3 snapshots (S1, S2, S3) committed. Now Flink source  ran initial scan again. For earliest strategy, S1 will be used as starting snapshot. For latest strategy, S3 will be used as starting snapshot. 
* case 2: there is 1 snapshot committed. Now both earliest and latest strategy will discover S1 as starting snapshot

regarding -1 snapshot id, is it a reserved special id? I wouldn't find the usage for snapshot id. Also, `IncrementalDataTableScan` seems to validate the start snapshot id. This may not work.
```
    Preconditions.checkArgument(
        snapshotIdsRange.contains(newFromSnapshotId),
        "from snapshot id %s not in existing snapshot ids range (%s, %s]",
        newFromSnapshotId, context().fromSnapshotId(), newToSnapshotId);
```

@yittg I definitely see your perspective, which is a valid interpretation of earliest and latest for empty table. In your interpretation, earliest and latest will behave the same (with some imaginary snapshot as starting point). all future snapshot should be discovered. To implement this model, we need to change `appendsBetween` implementation to support a special starting snapshot id, which is invalid and yet indicates a special starting point.

To me, both interpretations are reasonable. I also think for the practical purpose, how to interpret earliest and latest strategy for an empty table is not that important for streaming job. Most of the streaming job's bootstrap point is not really important, as it is arbitrary anyway.

Reply:
sounds reasonable to me, thanks.

Reply:
In a sync meeting with @rdblue , we agreed on two things
1. The starting snapshot should be inclusive. This is consistent with the behavior as Kafka source. It is also consistent with the empty table behavior below. I will rename the enum to `INCREMENTAL_FROM_...`, which hopefully clarify the inclusive behavior.
2. Earliest and latest start snapshot should work with empty table (all future snapshots should be included). This is original point from @yittg in this thread.

I looked into the IncrementalDataTableScan and SnapshotUtil. To achieve the above behavior, we can make fromSnapshotId nullable (not primitive) in TableScan . 
```java
TableScan appendsBetween(long->Long fromSnapshotId, long toSnapshotId);
```

1. We can just pass in starting snapshot's parent snapshotId as the fromSnapshotId  (could be null)
2. The earliest or latest start strategy for an empty table would just set the fromSnapshotId to null.

This method from SnapshotUtil already handles null fromSnapshotId
```java
private static List<Snapshot> snapshotsWithin(Table table, Long fromSnapshotId, long toSnapshotId)
```

**It is source compatible (as boxing happens automatically), but not binary compatible**. Is that acceptable?

cc @rdblue @openinx

Reply:
created a separate PR to support null fromSnapshotId for incremental scan.
https://github.com/apache/iceberg/pull/4529